functorch/functorch/experimental/cond.py

```diff
@@ -0,0 +1,139 @@
+import torch
+from torch._C import DispatchKey, DispatchKeySet, ExcludeDispatchKeyGuard
+from functorch.experimental.ops import PyOperator
+from torch.utils._pytree import tree_flatten
+from torch.fx.experimental.proxy_tensor import ProxyTensor, get_isolated_graphmodule, disable_proxy_modes_tracing
+import torch.utils._pytree as pytree
+from torch.dispatch.dispatcher import dispatcher_singleton
+
+"""
+We're going to define a `cond` operation.
+In order to do this, we need implementations for each of the dispatch keys.
+"""
+from contextlib import contextmanager
+
+# TODO(voz): Move out somewhere else once other py dispatched ops need it
+@contextmanager
+def suspend_mode(mode):
+    assert(mode is not None), "Cannot suspend None mode"
+    torch._C._set_torch_dispatch_mode(None)
+    try:
+        yield
+    finally:
+        torch._C._set_torch_dispatch_mode(mode)
+
+@contextmanager
+def enable_mode(mode):
+    curr_mode = torch._C._get_torch_dispatch_mode()
+    torch._C._set_torch_dispatch_mode(mode)
+    try:
+        yield
+    finally:
+        torch._C._set_torch_dispatch_mode(curr_mode)
+
+
+def trace_cond(proxy_mode, func_overload, args, kwargs=None):
+    assert kwargs is None or not kwargs
+    pred, true_fn, false_fn, operands = args
+
+    def _unwrap_proxy(e):
+        return e.proxy if isinstance(e, ProxyTensor) else e
+
+    def _extract_elements(e):
+        return e.elem if isinstance(e, ProxyTensor) else e
+
+    if isinstance(operands, ProxyTensor):
+        operands = [operands]  # Little hack because * on a single ProxyTensor unpacks it
+    else:
+        operands = operands
+
+    true_graph = get_isolated_graphmodule(true_fn, operands, {})
+    false_graph = get_isolated_graphmodule(false_fn, operands, {})
+    true_name = "true_graph"
+    false_name = "false_graph"
+    proxy_mode.tracer.root.register_module(true_name, true_graph)
+    proxy_mode.tracer.root.register_module(false_name, false_graph)
+
+    true_args = []
+    for node in true_graph.graph.nodes:
+        if node.op == 'output':
+            true_args.extend(*node.args)
+    
+    false_args = []
+    for node in false_graph.graph.nodes:
+        if node.op == 'output':
+            false_args.extend(*node.args)
+
+    assert(len(true_args) == len(false_args))
+    for i in range(len(true_args)):
+        t_arg_meta = true_args[i]
+        f_arg_meta = false_args[i]
+        # WIP don't look at this yet 
+        # assert(t_arg_meta.meta == f_arg_meta.meta)
+
+    if isinstance(operands, ProxyTensor):
+        operands = [operands]  # Prevent unwanted unpacking
+
+    args = (pred, true_graph, false_graph, operands)
+
+    proxy_args = pytree.tree_map(_unwrap_proxy, args)
+
+    elements = pytree.tree_map(_extract_elements, args)
+
+    # Does this need random slug appended so as not to collide?
+    proxy_res = proxy_mode.tracer.create_proxy('call_function', func_overload, proxy_args, kwargs,
```

Comment:
This is actually causing a problem today - it's storing the output not as a tensor, but as a `proxy(conditional)`.

The graph for this branch looks like:

```
def forward(self, flat_args):
    flat_args_1, flat_args_2, = fx_pytree.tree_flatten_spec([flat_args], self._in_spec)
    true_graph = self.true_graph
    false_graph = self.false_graph
    conditional = functorch_experimental_ops_cond(flat_args_2, true_graph, false_graph, [flat_args_1]);  flat_args_2 = true_graph = false_graph = flat_args_1 = None
    _tensor_constant0 = self._tensor_constant0
    add = _tensor_constant0.add(conditional);  _tensor_constant0 = conditional = None
    return pytree.tree_unflatten([add], self._out_spec)
```

And that isn't correct, or good. Piping conditional all the way back up *will* result in a tensor. However,  attempting to do anything with it (like this add operation here) is invalid. We need a way to both preserve this as a call_function (so that it can be dispatched at graph execution time, as opposed to trace time) AND have it affirm in the graph that the output of this call_function is undeniably a tensor, so that downstream ops can treat it as such. 

cc: @ezyang @Chillee @zou3519 @IvanYashchuk

Reply:
Resolved with pulling in @ezyang's diff to kill ProxyTensor.